core/src/main/scala/org/bitcoins/core/protocol/dlc/DLCTemplate.scala

```diff
@@ -0,0 +1,296 @@
+package org.bitcoins.core.protocol.dlc
+
+import org.bitcoins.core.currency._
+import org.bitcoins.core.protocol.dlc.ContractOraclePair._
+import org.bitcoins.core.protocol.dlc.DLCTemplate.validateMatchingOracleDescriptors
+import org.bitcoins.core.protocol.tlv._
+
+sealed trait DLCTemplate {
+  def totalCollateral: CurrencyUnit
+
+  def oracles: Vector[OracleAnnouncementTLV]
+
+  def oracleThreshold: Int
+
+  def contractDescriptor: ContractDescriptor
+
+  def oracleInfo: OracleInfo
+
+  def toContractInfo: ContractInfo
+
+  def individualCollateral: CurrencyUnit =
+    totalCollateral.satoshis / Satoshis(2)
+}
+
+sealed trait SingleOracleDLCTemplate extends DLCTemplate {
+  def oracle: OracleAnnouncementTLV
+
+  override def oracleInfo: SingleOracleInfo
+
+  final override def oracleThreshold: Int = 1
+  final override def oracles: Vector[OracleAnnouncementTLV] = Vector(oracle)
+}
+
+sealed trait MultiOracleDLCTemplate extends DLCTemplate {
+  override def oracleInfo: MultiOracleInfo[SingleOracleInfo]
+}
+
+sealed trait CFDTemplate extends DLCTemplate {
+  override def oracleInfo: NumericOracleInfo
+
+  def currentPrice: Long
+
+  def roundingIntervalsOpt: Option[RoundingIntervals]
+
+  override val contractDescriptor: NumericContractDescriptor = {
+    oracles.head.eventTLV.eventDescriptor match {
+      case _: EnumEventDescriptorV0TLV =>
+        throw new IllegalArgumentException(
+          "Cannot do a CFD with a EnumEventDescriptorV0TLV")
+      case decomp: DigitDecompositionEventDescriptorV0TLV =>
+        require(validateMatchingOracleDescriptors(oracles),
+                "Oracles must use same numDigits and base")
+
+        val func: Long => Long = { outcome =>
```

Comment:
this assumes the template holder is taking the fixed other position in the CFD right? Probably needs a bool so the other side can be used too right?